sycl/plugins/opencl/pi_opencl.cpp

```diff
@@ -1383,9 +1466,49 @@ pi_result piextKernelGetNativeHandle(pi_kernel kernel,
 // pi_level_zero.cpp for reference) Currently this is just a NOOP.
 pi_result piTearDown(void *PluginParameter) {
   (void)PluginParameter;
+  delete ExtFuncCaches;
+  ExtFuncCaches = nullptr;
   return PI_SUCCESS;
 }
 
+pi_result piContextRelease(pi_context Context) {
+#define RELEASE_EXT_FUNCS_CACHE_INTEL(t_pfx)                                   \
+  {                                                                            \
+    ExtFuncCache<t_pfx##Name, t_pfx##INTEL_fn> &Cache =                        \
+        ExtFuncCaches->get<t_pfx##Name, t_pfx##INTEL_fn>();                    \
+    std::lock_guard<std::mutex> CacheLock{Cache.Mtx};                          \
```

Comment:
This seems like an excessive amount of lookup and locking. I suppose we need to make sure all uses of a given function finishes prior to context-destruction, but would it maybe make sense to have a read-write lock or similar shared between them? Release would require write-access and the other functions would only need shared-access as they do not intend to change the cache. I am not sure if many spin-locks would be better in the end, but this appears like it is an overhead that happens in every context release, which would affect all programs, while calling extension functions is a little more work-dependent.